Allow command suggestions to autofill identifiers not under "minecraft" namespace

Ramixin opened the pull request:
# Issue
When typing commands, the autofill suggestions will only be able to autofill paths if they are under the `minecraft` namespace. This can be seen when typing `/give @s dirt` as the game will suggest `/give @s minecraft:dirt`. However, content not under the `minecraft` namespace needs to have its namespace typed before Minecraft autosuggests the full identifier. Typing in `/give @s cogwheel` will not have `create:cogwheel` suggested. The user has to type `/give @s create:` before Minecraft autosuggests the full identifier`.

# Change
This PR adds the mixin `CommandSourceMixin` to the `registry-sync-v0` module. The mixin allows Minecraft to autofill modded identifiers through cancelling the check that makes sure the autofill has the namespace `minecraft`.

Ramixin committed:
allows command suggestions to autofill identifiers not under the namespace "minecraft"

Ramixin committed:
Moved mixin to command module

Ramixin commented:
Moved the mixin to the `command-api-v2` module

chyzman commented:
https://github.com/chyzman/dontDoThat/blob/main/src/main/java/com/chyzman/dontdothat/mixin/CommandSourceMixin.java
does the same thing I'm not sure which method of doing this is better someone whose better at mixin might wanna compare them?

Ramixin commented:
@chyzman , I asked in the Mixin-dev channel and Llama said both work equally

TheTrueYak committed:
fix mixin not being an interface

because SOMEONE didn't even test the code after sniping it

chyzman reviewed (approved):
👍

Earthcomputer commented:
This should be pretty easy to add a unit test for?

modmuss50 commented:
Do we understand why this check is here in the first place, it looks very purposful and removing it without fully understanding the context worries me.

Have you got an example of what commands this affects, are there any behaviour changes when no mods are installed?

I worry that this PR has some unintended side affects that we may not want, its not something that we tend to change like this without careful consideration. As earth said a unit tests should be possible to add as well.

modmuss50 commented:
Please fix the build failuires.

modmuss50 commented:
Another question:

What is the behaviour when you have 2 identifers with the same path.

E.g you have both a `minecraft:dirt` and a `mymod:dirt` is there a worry that doing `/give @a dirt` will give you the wrong item? Will dirt be shown twice in the list of suggestions? This can also apply when two mods register something with the same path, it doesnt necessarily need to clash with `minecraft`.

When creating identifiers in code a string without a namespace always defaults to `minecraft` if I understand this right it changes that behaviour leading to possibly confusing and undefined behaviour.

chyzman commented:
This does not effect namespaceless identifiers, this does not change that

modmuss50 commented:
> This does not effect namespaceless identifiers, this does not change that

What does it change, sorry its not very clear to me what this PR does based on the description.

Earthcomputer commented:
> What does it change, sorry its not very clear to me what this PR does based on the description.

It changes auto-completion. If you type `dirt` and press tab, both `minecraft:dirt` and `mymod:dirt` will show up.

chyzman commented:
Minecraft namespaced identifiers will still appear before other namespaces tho

Ramixin committed:
Fixed formatting and added unit tests

Ramixin commented:
I have updated the mixin to be correctly formatted and I added two unit tests: one for testing if the non-vanilla identifier is suggested at all and the other checking that vanilla identifiers still have precedence.

Ramixin commented:
> Another question:
> 
> What is the behaviour when you have 2 identifers with the same path.
> 
> E.g you have both a `minecraft:dirt` and a `mymod:dirt` is there a worry that doing `/give @a dirt` will give you the wrong item? Will dirt be shown twice in the list of suggestions? This can also apply when two mods register something with the same path, it doesnt necessarily need to clash with `minecraft`.
> 
> When creating identifiers in code a string without a namespace always defaults to `minecraft` if I understand this right it changes that behaviour leading to possibly confusing and undefined behaviour.

When two identifiers have the same path, the vanilla identifier will always have precedence. So `minecraft:dirt` will always come before `mymod:dirt` in the suggestion list.

Identifiers without namespaces will always default to the vanilla namespace, so `/give @s dirt` will give `minecraft:dirt`. The namespace is still necessary on non-vanilla identifiers. The only difference is that the namespace can be autofilled.

Ramixin commented:
> Have you got an example of what commands this affects, are there any behaviour changes when no mods are installed?

This will affect any command that has an argument that uses identifiers to specify an object. (ex. `/give`, `/function`, `/effect`)
When no mods or datapacks are present, vanilla behavior is upheld since all identifiers are vanilla and would return `true` on the check anyway.

TheTrueYak commented:
yes can confirm. this PR does not at all change command syntax or logic itself, it just allows nonvanilla namespaces to appear as suggestions alongside vanilla ones; though the vanilla ones are still prioritized, as ramixin said

modmuss50 commented on fabric-command-api-v2/src/client/java/net/fabricmc/fabric/mixin/command/client/CommandSourceMixin.java:
```diff
@@ -0,0 +1,22 @@
+public interface CommandSourceMixin {
+	// Minecraft is hardcoded to only autofill identifiers with the "minecraft" namespace.
+	// This allows non-vanilla identifiers to be autofilled
+	@WrapOperation(
```
A Redirect is prefered over none conditional WrapOperation's

LlamaLad7 replied:
Not always. The sole purpose of a Redirect is to cause hard conflicts, there is no reason for it here. Anyone else changing this would likely change it the same way, and even if someone else wanted to customise the check to add their own modid for example, returning true always is perfectly compatible with that.

Earthcomputer replied:
Even if not `Redirect`, then `ModifyExpressionValue` could be used here instead.

cputnam-a11y replied:
I do think ModifyExpressionValue would be better in this case though, as it allows others code to still run around the checks, even though their value is discarded 
Oops, GitHub didn't update

chyzman replied:
[so this?](https://github.com/chyzman/dontDoThat/blob/main/src/main/java/com/chyzman/dontdothat/mixin/CommandSourceMixin.java)

Ramixin replied:
I chose `WrapOperation` over `ModifyExpressionValue` because it returns true no matter what, so skipping the other injections doesn't affect the result. But, if `ModifyExpressionValue` is preferred, I can switch it.

Earthcomputer replied:
Yes please use MEV

modmuss50 commented on fabric-command-api-v2/src/client/resources/fabric-command-api-v2.client.mixins.json:
```diff
@@ -3,6 +3,7 @@
   "package": "net.fabricmc.fabric.mixin.command.client",
   "compatibilityLevel": "JAVA_21",
   "client": [
+    "CommandSourceMixin",
```
Sort alpahatically.

Patbox replied:
Should this mixin be actually moved to common? Modified code can run on both sides the change doesn't require anything client only

Ramixin replied:
I made it client-side since I was assuming that this would only be useful for the client command auto-complete suggestions, but I have realized that there are some mods that utilize this in the terminal in some ways.

TheTrueYak replied:
probably. the only reason it was in client was because the initial mod was client side

Earthcomputer commented:
The build is still failing, you're missing license headers and import order. Try running `./gradlew spotlessApply`, it should fix all these issues.

Ramixin committed:
moved mixin to common and fixed style

Ramixin committed:
added license headers to files

Ramixin committed:
Fixed mixin location and changed injection to MEV

modmuss50 commented:
Checkstyle still doesnt pass, Ill fix it myself. Otherwise this is good to merge.

modmuss50 committed:
Fix checkstyle

JT122406 commented:
Can this be backported to 1.21.1?
